compiler/rustc_typeck/src/check/fn_ctxt.rs

```diff
@@ -1094,9 +1094,30 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 );
             }
 
-            if let Some(def_s) = def_span.map(|sp| tcx.sess.source_map().guess_head_span(sp)) {
-                err.span_label(def_s, "defined here");
+            if let Some(def_id) = def_id {
+                if let Some(node) = tcx.hir().get_if_local(def_id) {
+                    let mut spans = Vec::with_capacity(1 + expected_count);
+
+                    // Fall back to the full span if the node doesn't have an ident (tuple variants).
+                    spans.push(
+                        node.ident()
+                            .map(|ident| ident.span)
+                            .unwrap_or_else(|| tcx.hir().span(node.hir_id().unwrap())),
+                    );
+
+                    if let Some(id) = node.body_id() {
+                        let body = tcx.hir().body(id);
+                        spans.extend(body.params.iter().map(|param| param.span));
+                    }
+
+                    let def_kind = tcx.def_kind(def_id);
+                    err.span_note(
+                        MultiSpan::from_spans(spans),
```

Comment:
I would tweak this slightly in the following way

```suggestion
                    let mut spans: MultiSpan = node.ident()
                            .map(|ident| ident.span)
                            .unwrap_or_else(|| tcx.hir().span(node.hir_id().unwrap()))
                            .into();

                    if let Some(id) = node.body_id() {
                        let body = tcx.hir().body(id);
                        for param in body.params {
                            spans.push_span_label(param.span, "");
                        }
                    }

                    let def_kind = tcx.def_kind(def_id);
                    err.span_note(
                        spans,
```

so that the output looks more like

```
note: function defined here
  --> $DIR/E0061.rs:1:4
   |
LL | fn f(a: u16, b: &str) {}
   |    ^ ------- --------
```

Reply:
Done.